tools/codegen/selective_build/selector.py

```diff
@@ -0,0 +1,148 @@
+            assert isinstance(di_list, list)
+
+            debug_info = list(map(lambda x: str(x), di_list))
+
+        operators = {}
+        operators_dict = data.get('operators', {})
+        assert isinstance(operators_dict, dict)
+
+        for (k, v) in operators_dict.items():
+            operators[k] = SelectiveBuildOperator.from_yaml_dict(k, v)
+        return SelectiveBuilder(include_all_operators, debug_info, operators)
+
+    @staticmethod
+    def from_yaml_str(config_contents: str) -> 'SelectiveBuilder':
+        contents = yaml.load(config_contents)
+        return SelectiveBuilder.from_yaml_dict(contents)
+
+    @staticmethod
+    def from_yaml_path(config_path: str) -> 'SelectiveBuilder':
+        with open(config_path, 'r') as f:
+            contents = yaml.load(f)
+            return SelectiveBuilder.from_yaml_dict(contents)
+
+    @staticmethod
+    def from_legacy_op_registration_allow_list(
+            allow_list: Set[str],
+            is_root_operator: bool,
+            is_used_for_training: bool) -> 'SelectiveBuilder':
+        operators = {}
+        for op in allow_list:
+            operators[op] = {
+                'name': op,
+                'is_root_operator': is_root_operator,
+                'is_used_for_training': is_used_for_training,
+                'include_all_overloads': True,
+            }
+        return SelectiveBuilder.from_yaml_dict({
+            'operators': operators,
+        })
+
+    def is_operator_selected(self, name: str) -> bool:
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            return True
+        name = strip_operator_overload_name(name)
+        return name in self.operators and self.operators[name].include_all_overloads
+
+    def is_operator_selected_for_training(self, name: str) -> bool:
+        if not self.is_operator_selected(name):
+            return False
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            op: SelectiveBuildOperator = self.operators[name]
+            return op.is_used_for_training
+        name = strip_operator_overload_name(name)
+        if name not in self.operators:
+            return False
+        base_op: SelectiveBuildOperator = self.operators[name]
+        return base_op.include_all_overloads and base_op.is_used_for_training
+
+    def is_root_operator(self, name: str) -> bool:
+        if not self.is_operator_selected(name):
+            return False
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            op: SelectiveBuildOperator = self.operators[name]
+            return op.is_root_operator
+        name = strip_operator_overload_name(name)
+        if name not in self.operators:
+            return False
+        base_op: SelectiveBuildOperator = self.operators[name]
+        return base_op.include_all_overloads and base_op.is_root_operator
```

Comment:
root ops should always include overload part, right? We shouldn't apply `include_all_overloads` wildcard here, should we?

I know it breaks the symmetry - we can add a comment (include_all_overloads is being deprecated anyway).

Reply:
This one was a bit trick to reason about. Here's how I am thinking about it - which may be wrong.

Consider you have the following ops `aten::add`, `aten::add.int`, `aten::add.float`.
Now, the presence of `aten::add` could be 2 things.
1. The specific operator named `aten::add`, or
2. All overloads with `aten::add` as the base operator name

We have no way to distinguish between them except for the `include_all_overloads` flag. If `include_all_overloads=True`, then it means that `aten::add` stands for the family of operators whose base name is `aten::add`, whereas if `include_all_overloads=False`, then it means that this entry stands for exactly one operator that has the exact name `aten::add`.

In selective build, if `aten::add` is present with `include_all_overloads=False` and the question is asked about including `aten::add.double`, and that entry doesn't exist, we will fall back to checking if the base operator name `aten::add` exists. If it does, we need to additionally check if that operator name is [1] or [2] (above).